ee/control/control.go

```diff
@@ -0,0 +1,142 @@
+package control
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+// ControlService is the main object that manages the control service. It is responsible for fetching
+// and caching control data, and updating consumers and subscribers.
+type ControlService struct {
+	logger          log.Logger
+	cancel          context.CancelFunc
+	requestInterval time.Duration
+	consumers       map[string]consumer
+	subscribers     map[string][]subscriber
+	data            dataProvider
+	lastFetched     map[string]string
+}
+
+// consumer is an interface for something that consumes control server data updates.
+type consumer interface {
+	Update(io.Reader)
+}
+
+// subscriber is an interface for something that wants to be notified when a subsystem has been updated.
+type subscriber interface {
+	Ping()
+}
+
+// dataProvider is an interface for something that can retrieve control data. Authentication, HTTP,
+// file system access, etc. lives below this abstraction layer.
+type dataProvider interface {
+	Get(subsystem, cachedETag string) (etag string, data io.Reader, err error)
```

Comment:
I haven't read all the code yet, but how would this represent that there was nothing changed from the GET request?

Reply:
I've simplified this interface to remove the etag concept and is now just

```
Get(hash string) (data io.Reader, err error)
```

Instead of worrying about etags, the client can keep track of what object hashes it has asked for. When it finds the same hash for a subsystem, it can just choose not to request that same hash again. We also avoid hitting the server endpoint for it just to say "you already have this".